applications/HDF5Application/custom_io/hdf5_nd_data_io.cpp

```
@@ -0,0 +1,144 @@
+    } else {
+        // check if the partition table is written as a common table for set of datasets
+        // such as in the Variable data writing. This common data set should be having a partition table in the same
+        // group.
+        const auto group_name_pos = dataset_path.rfind("/");
+        const auto& partition_path = dataset_path.substr(0, group_name_pos + 1);
+
+        if (HasPartitionTable(*mpFile, partition_path)) {
+            std::tie(start_index, block_size) = StartIndexAndBlockSize(*mpFile, partition_path);
+        } else {
+            KRATOS_ERROR << "Partition table not found for dataset at \""
+                         << dataset_path << "\" either at the dataset location or at \""
+                         << partition_path << "\".";
+        }
+    }
+
+    // get raw data dimensions of h5.
+    const auto& h5_dimensions = mpFile->GetDataDimensions(dataset_path);
+    DenseVector<unsigned int> tensor_shape(h5_dimensions.size());
+    tensor_shape[0] = block_size;
+    std::copy(h5_dimensions.begin() + 1, h5_dimensions.end(), tensor_shape.data().begin() + 1);
+
+    NDDataPointerType p_nd_data;
+    if (mpFile->HasDataType<unsigned char>(dataset_path)) {
+        p_nd_data = Kratos::make_shared<NDData<unsigned char>>(tensor_shape);
+    } else if (mpFile->HasDataType<int>(dataset_path)) {
+        p_nd_data = Kratos::make_shared<NDData<int>>(tensor_shape);
+    } else if (mpFile->HasDataType<double>(dataset_path)) {
+        p_nd_data = Kratos::make_shared<NDData<double>>(tensor_shape);
+    } else {
+        KRATOS_ERROR
+                << "Unsupported data set type found at \"" << dataset_path
+                << "\". NDDatas only support uchar, bool, int and double data types.\n";
+    }
+
+    std::visit([this, &dataset_path, &tensor_shape, start_index](auto p_nd_data) {
+        if constexpr(!std::is_same_v<BareType<decltype(*p_nd_data)>, NDData<bool>>) {
```

wat? What happens if `NDData<bool>`? Nothing gets read?

---

The HDF5 always stores bools as `u_char`, so there is no possibility to have a `NDData<bool>`.. Even if you write an `NDData<bool>`, then it will be written to hdf5 as `unsigned char`. So when reading back, it will be read as an `u_char`.

Unfortunately, there is no way we can do the conversion, unelss we write an attribute stating the type of the nd array. (Which I don't like at all).

At the moment, we don't have any `TensorAdaptor<bool>` scenarios as well.

---

Then I'd throw an exception for now.

---

The error is already there at the following
https://github.com/KratosMultiphysics/Kratos/blob/c31feed93f2bad0ad9eae9726843294b6e2570f6/applications/HDF5Application/custom_io/hdf5_nd_data_io.cpp#L122-L133